dipy/denoise/localpca.py

```diff
@@ -0,0 +1,137 @@
+
+    # declare arrays for theta and thetax
+    theta = np.zeros(arr.shape, dtype=np.float64)
+    thetax = np.zeros(arr.shape, dtype=np.float64)
+
+    for k in range(patch_radius, arr.shape[2] - patch_radius):
+        for j in range(patch_radius, arr.shape[1] - patch_radius):
+            for i in range(patch_radius, arr.shape[0] - patch_radius):
+
+                X = np.zeros(
+                    (patch_size * patch_size * patch_size, arr.shape[3]))
+                M = np.zeros(arr.shape[3])
+
+                temp = arr[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1, :]
+                X = temp.reshape(
+                    patch_size *
+                    patch_size *
+                    patch_size,
+                    arr.shape[3])
+                # compute the mean and normalize
+                M = np.mean(X, axis=0)
+                X = X - M
+                # PCA. There are two ways to do this:
+                # # 1. Compute the covariance matrix C = X_transpose X
+                # C = np.dot(X.T, X)
+                # C = C / X.shape[0]
+                # # and then decompose it:
+                # d, W = sla.eigh(C)
+
+                # 2. Alternatively, calculate the eigenvalues and
+                #  eigenvectors of the covariance matrix through an SVD:
+                U, S, Vt = svd(X, *svd_args)[:3]
+                # Items in S are the eigenvalues, but in ascending order
+                # We invert the order (=> descending), square and normalize
+                # \lambda_i = s_i^2 / n
+                d = S[::-1] ** 2 / X.shape[0]
+                # Rows of Vt are eigenvectors, but also in ascending eigenvalue
+                # order:
+                W = Vt[::-1].T
+                d[d < tau[i, j, k, :]] = 0
+                W_hat = np.zeros_like(W)
+                W_hat[:, d > 0] = W[:, d > 0]
+                Y = X.dot(W_hat)
+                X_est = Y.dot(np.transpose(W_hat))
+
+                # theta value
+                temp = X_est + \
+                    np.array([M, ] * X_est.shape[0], dtype=np.float64)
+                temp = temp.reshape(
+                    patch_size, patch_size, patch_size, arr.shape[3])
+                # Also update the estimate matrix which is X_est * theta
+
+                theta[i - patch_radius: i + patch_radius + 1,
+                      j - patch_radius: j + patch_radius + 1,
+                      k - patch_radius: k + patch_radius + 1,
+                      :] = theta[i - patch_radius: i + patch_radius + 1,
+                                 j - patch_radius: j + patch_radius + 1,
+                                 k - patch_radius: k + patch_radius + 1,
+                                 :] + 1.0 / (1.0 + np.linalg.norm(d,
+                                                                  ord=0))
```

Comment:
Not sure if this is correct!
weights have to be equal to 1/(1 + nd) where nd is the number of non-zero eigenvalues.
np.linalg.norm is doing the sum of non-zero eigenvalues instead. In the original article the used the norm of matrix d, however matrix d is previously converted to zeros and ones.